xla/service/gpu/autotuning/gemm_fusion_autotuner_cuda.cc

```diff
@@ -0,0 +1,383 @@
+/* Copyright 2024 The OpenXLA Authors.
```

Comment:
I looked at the diff between this file and the `_rocm` version. At the moment it is very small and I think we'll be much better off if we keep this a single file with a few places where the code branches. Also in general, I suspect that's desirable for the long-term.

Do you have a strong reason why splitting this code and duplicating a lot of it is desirable? If not, please revert this change.

Reply:
The reason for creating two files (_cuda and _rocm versions) is that there are parts of the code that cannot be compiled as a part of cuda or rocm build (e.g. missing cuda header files in rocm build).
Keeping it in single file would require introduction of #if GOOGLE_CUDA (or similar) guards.
My understanding is that introduction of these new #ifdef is not acceptable,

Reply:
You are correct, we should not introduce new `#ifdef` macros. However, my comment still mostly stands. We need a hybrid solution:

- Leave all common code in the current file.
- Isolate only parts that require Rocm/Cud headers into the new `_rocm`, `_cuda` files. If I look at the diff between the two, I believe that's just:
  - `const int64_t BLAS_GEMM_DEFAULT`
  - `GetCuDnnPlanCount` and more generally the CuDNN bit of code.
- Everything else could stay in one file. E.g. for `GetDefaultTritonConfigs()` the code can just add the common configs unconditionally, then check for cuda and add the remaining bits.

@beckerhe For having two different implementations of a common header (`rocm` and `cuda`), is the current solution OK (different `srcs` based on `if_rocm/cuda_configured`)? Is it preferable to have two entirely separate targets that are then conditionally added to `deps`, e.g. like we do [here](https://github.com/openxla/xla/blob/main/xla/mlir/tools/mlir_interpreter/dialects/BUILD#L73-L110)?

Reply:
> Is it preferable to have two entirely separate targets that are then conditionally added to deps, e.g. like we do [here](https://github.com/openxla/xla/blob/main/xla/mlir/tools/mlir_interpreter/dialects/BUILD#L73-L110)?

Yes, that would be preferrable for a couple of reasons:

1. Both targets can have different dependencies and different build flags if needed
2. Dependency cleaners like `build_cleaner` only work if these are separate targets